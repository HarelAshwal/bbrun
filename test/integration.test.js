function run(options) {
  const exec = require("shelljs").exec;
  return exec(`node index.js ${options}`, {
    silent: true
  });
}

const PWD = require("shelljs").pwd();

const norm = input => input.replace(new RegExp(PWD, "g"), "PWD");

describe("single step pipeline", () => {
  it("executes the default step with no arguments", () => {
    const res = run(
      "--template test/templates/pipeline-one-step.yml --dry-run"
    );
    expect(norm(res.stdout)).toMatchSnapshot();
    expect(res.stderr).toBe("");
    expect(res.code).toBe(0);
  });

  it("executes the step by name", () => {
    const res = run(
      "test --template test/templates/pipeline-one-step.yml --dry-run"
    );
    expect(norm(res.stdout)).toMatchSnapshot();
    expect(res.stderr).toBe("");
    expect(res.code).toBe(0);
  });

  it("fails executing non-existing name", () => {
    const res = run(
      "foo --template test/templates/pipeline-one-step.yml --dry-run"
    );
    expect(norm(res.stdout)).toBe("");
    expect(res.stderr).toBe('couldn\'t find step with name="foo"\n');
    expect(res.code).toBe(1);
  });
});

it("invalid template should fail", () => {
  const res = run("--template test/templates/invalid-template.yml --dry-run");
  expect(norm(res.stdout)).toBe("");
  expect(res.stderr).toMatchSnapshot();
  expect(res.code).toBe(1);
});

it("no image template should use default atlassian image", () => {
  const res = run("--template test/templates/no-image-template.yml --dry-run");
  expect(norm(res.stdout)).toMatchSnapshot();
  expect(res.stderr).toBe("");
  expect(res.code).toBe(0);
});

describe("template with multiple steps in the default pipeline", () => {
  it("should execute all the default steps when no argument is provided", () => {
    const res = run(
      "--template test/templates/pipeline-multiple-steps.yml --dry-run"
    );
    expect(norm(res.stdout)).toMatchSnapshot();
    expect(res.stderr).toBe("");
    expect(res.code).toBe(0);
  });
  it("should execute a single step by name", () => {
    const res = run(
      "step2 --template test/templates/pipeline-multiple-steps.yml --dry-run"
    );
    expect(norm(res.stdout)).toMatchSnapshot();
    expect(res.stderr).toBe("");
    expect(res.code).toBe(0);
  });
  it("should execute a single step by name with spaces in the name", () => {
    const res = run(
      '"Step Three" --template test/templates/pipeline-multiple-steps.yml --dry-run'
    );
    expect(norm(res.stdout)).toMatchSnapshot();
    expect(res.stderr).toBe("");
    expect(res.code).toBe(0);
  });
});

it("template with no default should fail when no step name is passed", () => {
  const res = run(
    "--template test/templates/no-default-pipeline.yml --dry-run"
  );
  expect(norm(res.stdout)).toBe("");
  expect(res.stderr).toBe("default pipeline not found\n");
  expect(res.code).toBe(1);
});
