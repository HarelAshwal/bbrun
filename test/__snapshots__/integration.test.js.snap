// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid template should fail 1`] = `
"Malformed template, check https://confluence.atlassian.com/bitbucket/configure-bitbucket-pipelines-yml-792298910.html
"
`;

exports[`no image template should use default atlassian image 1`] = `
"executing step in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"testing stuff\\"
"
`;

exports[`single step pipeline executes the default step with no arguments 1`] = `
"executing step in \\"ubuntu\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws ubuntu chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"testing stuff\\"
"
`;

exports[`single step pipeline executes the step by name 1`] = `
"executing step \\"test\\" in \\"ubuntu\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws ubuntu chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"testing stuff\\"
"
`;

exports[`template with multiple steps in the default pipeline should execute a named step in a branch 1`] = `
"executing step \\"master_step_2\\" in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"master step 2\\"
"
`;

exports[`template with multiple steps in the default pipeline should execute a single step by name 1`] = `
"executing step \\"step2\\" in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"step2\\"
"
`;

exports[`template with multiple steps in the default pipeline should execute a single step by name with spaces in the name 1`] = `
"executing step \\"Step Three\\" in \\"ubuntu\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws ubuntu chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"step3\\"
"
`;

exports[`template with multiple steps in the default pipeline should execute all the default steps when no argument is provided 1`] = `
"executing step in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"testing stuff\\"
executing step in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"step2\\"
executing step in \\"ubuntu\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws ubuntu chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"step3\\"
"
`;

exports[`template with multiple steps in the default pipeline should execute all the steps in a branch if no step provided 1`] = `
"executing step in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"testing stuff in master branch\\"
executing step in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"master step 2\\"
"
`;

exports[`template with multiple steps in the default pipeline should understand asterisk branches 1`] = `
"executing step \\"branch step 2\\" in \\"atlassian/default-image:latest\\"
docker command:
	docker run  -P -v PWD:/ws -w /ws atlassian/default-image:latest chmod +x .bbrun.sh && ./.bbrun.sh
build script:
	set -e
	echo \\"testing stuff in any branch\\"
"
`;
